### DISCUSSION TITLE
How to develop plugins in APISIX?

### DISCUSSION THREAD

**Discussion by zgLiuBing** [apache/apisix#5272] [Q&A] [answered]
### Improved or typo fixed.

插件的执行阶段都有哪些啊，文档上就两个啊 rewrite 和 access 阶段，我需要数据返回的阶段

### Solution

插件的执行阶段都有哪些啊，文档上就两个啊 rewrite 和 access 阶段，我需要数据返回的阶段

**Comment by zgLiuBing** [comment]
我在开发一个脱敏插件，需要对返回的插件做处理

**Comment by shuaijinchao** [comment, marked as answer]
You need to understand the concept of OpenResty's execution phase. The upstream response data can be obtained through `header_filter` and `body_filter`. The APISIX plug-in also supports the definition of these two stages.
https://moonbingbing.gitbooks.io/openresty-best-practices/content/ngx_lua/phase.html

It is recommended to read this book, which can help you solve many basic problems in the development of OpenResty.
https://moonbingbing.gitbooks.io/openresty-best-practices/content/
